flashlight/fl/tensor/backend/jit/opt/backends/onednn/OneDnnOpFusion.cpp

```diff
@@ -0,0 +1,240 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include "flashlight/fl/tensor/backend/jit/opt/backends/onednn/OneDnnOpFusion.h"
+
+#include "flashlight/fl/tensor/backend/jit/ir/BinaryNode.h"
+#include "flashlight/fl/tensor/backend/jit/ir/CustomNode.h"
+#include "flashlight/fl/tensor/backend/onednn/OneDnnBackend.h"
+#include "flashlight/fl/tensor/backend/onednn/OneDnnTensor.h"
+#include "flashlight/fl/tensor/backend/onednn/Utils.h"
+
+#include "dnnl.hpp"
+
+namespace fl {
+
+struct BinopInfo {
+  Node* rhsNode;
+  BinaryOp op;
+};
+
+struct OneDnnOpFusion::SearchState {
+  Node* searchRoot;
+  // Assume `searchRoot == binop2`
+  //
+  // x0  x1
+  //  \  /
+  //  binop1  x2
+  //     \  /
+  //    binop2
+  //
+  // accumulatedBinopInfos: { { binop2->op(), x2 }, { binop1->op(), x1 } }
+  std::vector<BinopInfo> accumulatedBinopInfos;
+};
+
+namespace {
+
+// https://github.com/oneapi-src/oneDNN/blob/adeda9fcc20149effb1bffc051262810e9f3138c/include/oneapi/dnnl/dnnl_types.h#L2914
+static constexpr unsigned kOneDnnMaxNumPostOps = 32;
+
+dnnl::memory::data_type getOneDnnTypeWithLargestRange(
```

Comment:
nit -- might be good to detail this type promotion behavior somewhere

Reply:
Just added a documentation at the use site.

Eventually this will be replaced by a type inference mechanism in JIT.